libs/utils/src/logging.rs

```diff
@@ -330,37 +330,89 @@ impl std::fmt::Debug for SecretString {
 #[inline]
-pub async fn log_slow<O>(name: &str, threshold: Duration, f: impl Future<Output = O>) -> O {
-    // TODO: we unfortunately have to pin the future on the heap, since GetPage futures are huge and
-    // won't fit on the stack.
-    let mut f = Box::pin(f);
+pub async fn log_slow<F, O>(name: &str, threshold: Duration, f: std::pin::Pin<&mut F>) -> O
+where
+    F: Future<Output = O>,
+{
+    monitor_slow_future(
+        threshold,
+        threshold, // period = threshold
+        f,
+        |MonitorSlowFutureCallback {
+             ready,
+             elapsed_total,
+             elapsed_since_last_callback: _,
+         }| {
+            if ready {
+                // NB: we check if we exceeded the threshold even if the timeout never fired, because
+                // scheduling or execution delays may cause the future to succeed even if it exceeds the
+                // timeout. This costs an extra unconditional clock reading, but seems worth it to avoid
+                // false negatives.
+                if elapsed_total >= threshold {
+                    info!(
+                        "slow {name} completed after {:.3}s",
+                        elapsed_total.as_secs_f64()
+                    );
+                }
+            } else if elapsed_total > threshold {
+                info!(
+                    "slow {name} still running after {:.3}s",
+                    elapsed_total.as_secs_f64()
+                );
+            }
+        },
+    )
+    .await
+}
 
+/// Poll future `fut` to completion, invoking callback `cb` at the given `period` and once after the future completes.
+#[inline]
+pub async fn monitor_slow_future<F, O>(
+    threshold: Duration,
+    period: Duration,
+    mut fut: std::pin::Pin<&mut F>,
+    mut cb: impl FnMut(MonitorSlowFutureCallback),
+) -> O
+where
+    F: Future<Output = O>,
+{
     let started = Instant::now();
     let mut attempt = 1;
-
+    let mut last_cb = started;
     loop {
         // NB: use timeout_at() instead of timeout() to avoid an extra clock reading in the common
         // case where the timeout doesn't fire.
-        let deadline = started + attempt * threshold;
-        if let Ok(output) = tokio::time::timeout_at(deadline, &mut f).await {
-            // NB: we check if we exceeded the threshold even if the timeout never fired, because
-            // scheduling or execution delays may cause the future to succeed even if it exceeds the
-            // timeout. This costs an extra unconditional clock reading, but seems worth it to avoid
-            // false negatives.
-            let elapsed = started.elapsed();
-            if elapsed >= threshold {
-                info!("slow {name} completed after {:.3}s", elapsed.as_secs_f64());
-            }
+        let deadline = if attempt == 1 {
+            started + threshold
+        } else {
+            started + threshold + attempt * period
+        };
```

Comment:
I think this should unconditionally be `let deadline = started + threshold + (attempt - 1) * period`, otherwise we're waiting 2x period after the first attempt fails.